lib/src/merged_tree.rs

```diff
@@ -971,6 +974,235 @@ impl Iterator for TreeDiffIterator<'_> {
+            builder.build()
+        } else {
+            Merge::resolved(Tree::null(store.clone(), dir.clone()))
+        };
+        // Maintain the type of tree, so we resolve `TreeValue::Conflict` as necessary
+        // in the subtree
+        if legacy_format {
+            Ok(MergedTree::Legacy(trees.into_resolved().unwrap()))
+        } else {
+            Ok(MergedTree::Merge(trees))
+        }
+    }
+
+    fn add_dir_diff_items(
+        &mut self,
+        dir: RepoPath,
+        tree_diff: BackendResult<(MergedTree, MergedTree)>,
+    ) {
+        let (tree1, tree2) = match tree_diff {
+            Ok(trees) => trees,
+            Err(err) => {
+                self.items.insert(DiffStreamKey::normal(dir), Err(err));
+                return;
+            }
+        };
+
+        for basename in merged_tree_basenames(&tree1, &tree2) {
+            let value_before = tree1.value(basename);
+            let value_after = tree2.value(basename);
+            if value_after != value_before {
+                let path = dir.join(basename);
+                let before = value_before.to_merge();
+                let after = value_after.to_merge();
+                if (before.is_tree() || after.is_tree()) && !self.matcher.visit(&path).is_nothing()
+                {
+                    let before_tree_future = Self::tree(
+                        self.store.clone(),
+                        self.legacy_format_before,
+                        path.clone(),
+                        before.clone(),
+                    );
+                    let after_tree_future = Self::tree(
+                        self.store.clone(),
+                        self.legacy_format_after,
+                        path.clone(),
+                        after.clone(),
+                    );
+                    let both_trees_futures =
+                        async { futures::try_join!(before_tree_future, after_tree_future) };
+                    self.pending_trees
+                        .push_back((path.clone(), Box::pin(both_trees_futures)));
+                }
+                if self.matcher.matches(&path) {
+                    self.items
+                        .insert(DiffStreamKey::normal(path), Ok((before, after)));
+                }
+            }
+        }
+    }
+}
+
+impl Stream for TreeDiffStreamImpl<'_> {
+    type Item = (RepoPath, BackendResult<(MergedTreeValue, MergedTreeValue)>);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut need_items = false;
+        while !(self.items.is_empty() && self.pending_trees.is_empty()) {
+            // Go through all pending tree futures and poll them.
+            let mut pending_index = 0;
+            while pending_index < self.pending_trees.len()
+                && (need_items
+                    || pending_index < self.max_concurrent_reads
+                    || self.items.len() < self.max_queued_items)
+            {
```

Comment:
Just a question. I think the output can be emitted in different order if pending/queued items exceeds. Is my understanding correct?

Reply:
Not intentionally at least :) Did you find a case where it does emit items in a different order?

Reply:
I don't have a particular example. I just thought that would be possible because `pending_trees` and `items` are sorted differently, and `pending_trees` can be limited earlier by `max_concurrent_reads`?

Reply:
Ah, I was wrong. There's a pending `item` to block too early emission. And that's why we have inserted a tree to both `pending_trees` and `items`.